src/hotspot/share/ci/ciType.cpp

```diff
@@ -140,3 +140,16 @@ void ciReturnAddress::print_impl(outputStream* st) {
 ciReturnAddress* ciReturnAddress::make(int bci) {
   GUARDED_VM_ENTRY(return CURRENT_ENV->get_return_address(bci);)
 }
+
+ciWrapper::ciWrapper(ciType* type, int properties)
+    : ciType(type->basic_type()),
+      _type(type),
+      _properties(properties) {
+  assert(!type->is_wrapper(), "Thou shall not double wrap!");
+  assert(type->is_inlinetype()
+             // An abstract value type is an instance_klass
+             || (type->is_instance_klass() && type->as_instance_klass()->flags().is_abstract() && !type->as_instance_klass()->flags().is_identity())
```

Comment:
Just wondering, is the `type->as_instance_klass()->flags().is_abstract()` condition even needed? Same in `ciTypeFlow::get_start_state()`.

Reply:
Probably not. An instance class without identity must be an abstract value class. At least, I don't see another way... But just to be sure, I'm running some tests.

Reply:
Right, I think so too. Thanks for checking!

Reply:
I've realized that since `ciInlineKlass` is a class derived from `ciInstanceKlass`, `type->is_instance_klass()` is also true for `ciInlineKlass`, which messed up my experiment but I guess it's reasonable.